vowpalwabbit/feature_group.cc

```diff
@@ -48,6 +49,14 @@ void features::truncate_to(size_t i, float sum_feat_sq_of_removed_section)
   if (indicies.end() != indicies.begin()) { indicies.resize_but_with_stl_behavior(i); }
 
   if (space_names.size() > i) { space_names.erase(space_names.begin() + i, space_names.end()); }
+
+  while (!namespace_extents.empty() && namespace_extents.back().begin_index >= i) { namespace_extents.pop_back(); }
```

Comment:
as an alternative std::find_if and std::erase can be used here

Reply:
pop_back is used instead of a combo of remove_if and erase because pop_back is guaranteed to be constant time. If erase is used then it can be used in such a way to be linear over the size of the container. 

The idea is that removals *always* occur from the back and so if the situation cannot be implemented with pop_back then there is some change of semantics that goes against the design.

Reply:
I agree it's the same complexity, it's more about clearly stating we always remove from the back. This would make it a little less likely to be changed to some inner removal in future. The constructions of extents (should) ensure there are no empty extents in the middle and for an empty namespace there would be nothing to remove.